File: src/main/scala/com/flink/tutorials/scala/api/time/PeriodicWatermark.scala
package com.flink.tutorials.scala.api.time

import java.text.SimpleDateFormat

import org.apache.flink.streaming.api.TimeCharacteristic
import org.apache.flink.streaming.api.functions.{AssignerWithPeriodicWatermarks, AssignerWithPunctuatedWatermarks}
import org.apache.flink.streaming.api.functions.timestamps.BoundedOutOfOrdernessTimestampExtractor
import org.apache.flink.streaming.api.scala._
import org.apache.flink.streaming.api.watermark.Watermark
import org.apache.flink.streaming.api.windowing.time.Time

object PeriodicWatermark {

  def main(args: Array[String]): Unit = {

    val env = StreamExecutionEnvironment.getExecutionEnvironment
    env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime)

    // 每5000毫秒生成一个Watermark
    env.getConfig.setAutoWatermarkInterval(5000L)

    val socketSource = env.socketTextStream("localhost", 9000)

    val input = socketSource.map{
      line => {
        val arr = line.split(" ")
        val id = arr(0)
        val time = arr(1).toLong
        (id, time)
      }
    }

    val watermark = input.assignTimestampsAndWatermarks(new MyPeriodicAssigner)
    val boundedOutOfOrder = input.assignTimestampsAndWatermarks(
      new BoundedOutOfOrdernessTimestampExtractor[(String, Long)](Time.minutes(1)) {
        override def extractTimestamp(element: (String, Long)): Long = {
          element._2
        }
    })

    env.execute("periodic and punctuated watermark")

  }

  // 假设数据流的元素有两个字段(String, Long)，其中第二个字段是该元素的时间戳
  class MyPeriodicAssigner extends AssignerWithPeriodicWatermarks[(String, Long)] {
    val bound: Long = 60 * 1000     // 1分钟
    var maxTs: Long = Long.MinValue // 已抽取的Timestamp最大值

    val formatter = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS")

    override def extractTimestamp(element: (String, Long), previousElementTimestamp: Long): Long = {
      println("extractTimestamp is inovked for element " + element._1 + "@" + element._2)
      // 更新maxTs为当前遇到的最大值
      maxTs = maxTs.max(element._2)
      // 使用第二个字段作为这个元素的Event Time
      element._2
    }

    override def getCurrentWatermark: Watermark = {
      println("getCurrentWatermark method is invoked @:" + formatter.format(System.currentTimeMillis()))
      // Watermark比Timestamp最大值慢1分钟
      val watermark = new Watermark(maxTs - bound)
      println("watermark timestamp @:" + formatter.format(watermark.getTimestamp))
      watermark
    }
  }

  // 第二个字段是时间戳，第三个字段判断是否为Watermark的标记
  class MyPunctuatedAssigner extends AssignerWithPunctuatedWatermarks[(String, Long, Boolean)] {

    override def extractTimestamp(element: (String, Long, Boolean), previousElementTimestamp: Long): Long = {
      element._2
    }

    override def checkAndGetNextWatermark(element: (String, Long, Boolean), extractedTimestamp: Long): Watermark = {
      if (element._3)
        new Watermark(extractedTimestamp)
      else
        null
    }
  }

}

File: src/main/java/com/flink/tutorials/java/api/time/PeriodicWatermark.java
package com.flink.tutorials.java.api.time;

import com.flink.tutorials.java.api.utils.random.RandomId;
import com.flink.tutorials.java.api.utils.random.RandomSource;
import com.flink.tutorials.java.api.utils.stock.StockPrice;
import com.flink.tutorials.java.api.utils.stock.StockSource;
import org.apache.flink.api.common.typeinfo.Types;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.api.TimeCharacteristic;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.AssignerWithPeriodicWatermarks;
import org.apache.flink.streaming.api.watermark.Watermark;

public class PeriodicWatermark {

    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime);

        // 每5000毫秒生成一个Watermark
        env.getConfig().setAutoWatermarkInterval(5000L);
        DataStream<String> socketSource = env.socketTextStream("localhost", 9000);
        DataStream<Tuple2<String, Long>> input = socketSource.map(
                line -> {
                    String[] arr = line.split(" ");
                    String id = arr[0];
                    long time = Long.parseLong(arr[1]);
                    return Tuple2.of(id, time);
                })
                .returns(Types.TUPLE(Types.STRING, Types.LONG));

//        DataStream<> watermark = input.assignTimestampsAndWatermarks(new MyPeriodicAssigner)
//
//        DataStream<> watermark = input.assignTimestampsAndWatermarks(new MyPeriodicAssigner)
//        val boundedOutOfOrder = input.assignTimestampsAndWatermarks(
//                new BoundedOutOfOrdernessTimestampExtractor[(String, Long)](Time.minutes(1)) {
//            override def extractTimestamp(element: (String, Long)): Long = {
//                    element._2
//            }
//        })

        env.execute("periodic and punctuated watermark");

//        val socketSource = env.socketTextStream("localhost", 9000)
//
//        val input = socketSource.map{
//            line => {
//                val arr = line.split(" ")
//                val id = arr(0)
//                val time = arr(1).toLong
//                        (id, time)
//            }
//        }

        env.execute("event time and watermark");
    }

    public static class MyPeriodicAssigner implements AssignerWithPeriodicWatermarks<Tuple2<String, Long>> {

        private long bound = 60 * 1000;
        private long maxTs = Long.MIN_VALUE;

        @Override
        public long extractTimestamp(Tuple2<String, Long> element, long previousElementTimestamp) {
            // 更新maxTs为当前遇到的最大值
            maxTs = Math.max(maxTs, element.f1);
            return element.f1;
        }

        @Override
        public Watermark getCurrentWatermark() {
            // Watermark比Timestamp最大值慢1分钟
            Watermark watermark = new Watermark(maxTs - bound);
            return watermark;
        }
    }

}
